runtime/client_http_request.go

```diff
@@ -203,21 +212,76 @@ func (req *ClientHTTPRequest) Do() (*ClientHTTPResponse, error) {
 		/* coverage ignore next line */
 		return nil, err
 	}
+	var retryCount int64 = 1
+	var res *http.Response
+
+	//when timeoutAndRetryOptions per request is not configured, use default client level timeout
+	if req.timeoutAndRetryOptions == nil || req.timeoutAndRetryOptions.MaxAttempts == 0 {
+		res, err = req.client.Client.Do(req.httpReq.WithContext(ctx))
+	} else {
+		res, retryCount, err = req.executeDoWithRetry(ctx) //new code for retry and timeout per ep level
+	}
 
-	res, err := req.client.Client.Do(req.httpReq.WithContext(ctx))
 	span.Finish()
+
 	if err != nil {
-		req.ContextLogger.ErrorZ(req.ctx, "Could not make outbound request", zap.Error(err))
-		return nil, err
+		req.ContextLogger.ErrorZ(req.ctx, fmt.Sprintf("Could not make http outbound %s.%s request",
+			req.ClientID, req.MethodName), zap.Error(err))
+		return nil, errors.Wrapf(err, "errors while making outbound %s.%s request", req.ClientID, req.MethodName)
 	}
 
 	// emit metrics
-	req.Metrics.IncCounter(req.ctx, clientRequest, 1)
+	req.Metrics.IncCounter(req.ctx, clientRequest, retryCount)
 
 	req.res.setRawHTTPResponse(res)
 	return req.res, nil
 }
 
+//executeDoWithRetry will execute executeDo with retries
+func (req *ClientHTTPRequest) executeDoWithRetry(ctx context.Context) (*http.Response, int64, error) {
+	var err error
+	var res *http.Response
+	var retryCount int64 = 0
+
+	for i := 0; i < req.timeoutAndRetryOptions.MaxAttempts; i++ {
+		retryCount++
+		res, err = req.executeDo(ctx)
+
+		if err == nil {
+			return res, retryCount, nil
+		}
+
+		// if attempts are pending, wait for backoff duration before next attempt
+		if i+1 < req.timeoutAndRetryOptions.MaxAttempts {
+			timer := time.NewTimer(req.timeoutAndRetryOptions.BackOffTimeAcrossRetriesInMs)
+			select {
+			case <-timer.C:
+			}
+		}
+
+		req.ContextLogger.GetLogger().Info("errors while making http outbound request",
+			zap.Error(err),
+			zap.String("clientId", req.ClientID), zap.String("methodName", req.MethodName),
+			zap.String("attempt", string(rune(retryCount))),
+			zap.String("maxAttempts", string(rune(req.timeoutAndRetryOptions.MaxAttempts))))
+
+		//TODO (future releases) - make retry conditional, inspect error/response and then retry
```

Comment:
Cany maybe model as retry strategy and use strategy pattern to have various types of strategies. Will be extensible

Reply:
+1

Additionally Run this with `goleak` to make sure there are no leaks

https://github.com/uber-go/goleak

Reply:
Added retry check that can be extended with a new strategy in future.

Ran test cases with **goleak** no leakage due to this change